File: tests/djvCorePyTest/AnimationTest.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2020 Darby Johnston
# All rights reserved.

import djvCorePy as d
import djvCmdLineAppPy as c

import sys
import unittest
from datetime import timedelta

class AnimationTest(unittest.TestCase):

    app = None

    def animationValue(self, v):
        buf = ""
        for i in range(0, int(v * 40.0)):
            buf += "#"
        print(buf)
    def animationFinished(self, v):
        print(v)
        self.app.exit()

    def test_Animation(self):
        for i in [
            d.Animation.Type.Linear,
            d.Animation.Type.EaseIn,
            d.Animation.Type.EaseOut,
            d.Animation.Type.SmoothStep,
            d.Animation.Type.Sine]:
            self.app = c.Application.create(sys.argv)
            a = d.Animation.Animation.create(self.app)
            self.assertFalse(a.isActive())
            self.assertFalse(a.isRepeating())
            a.setType(i)
            a.setRepeating(True)
            self.assertEqual(a.getType(), i)
            self.assertTrue(a.isRepeating())
            a.start(
                0.0,
                1.0,
                timedelta(seconds=1),
                self.animationValue,
                self.animationFinished)
            self.app.run()

if __name__ == '__main__':
    unittest.main()

File: tests/djvSystemTest/AnimationTest.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2004-2020 Darby Johnston
// All rights reserved.

#include <djvSystemTest/AnimationTest.h>

#include <djvSystem/Animation.h>
#include <djvSystem/AnimationFunc.h>
#include <djvSystem/Context.h>

#include <sstream>

using namespace djv::Core;
using namespace djv::System;

namespace djv
{
    namespace SystemTest
    {
        AnimationTest::AnimationTest(
            const File::Path& tempPath,
            const std::shared_ptr<Context>& context) :
            ITickTest("djv::SystemTest::AnimationTest", tempPath, context)
        {}
        
        void AnimationTest::run()
        {
            if (auto context = getContext().lock())
            {
                for (auto i : Animation::getTypeEnums())
                {
                    auto animation = Animation::Animation::create(context);
                    DJV_ASSERT(!animation->isActive());
                    DJV_ASSERT(!animation->isRepeating());
                    animation->setType(i);
                    animation->setRepeating(true);
                    DJV_ASSERT(animation->getType() == i);
                    DJV_ASSERT(animation->isRepeating());
                    
                    animation->start(
                        0.F,
                        1.F,
                        std::chrono::milliseconds(250),
                        [this](float v)
                        {
                            std::stringstream ss;
                            ss << v;
                            _print(ss.str());
                        },
                        [this](float v)
                        {
                            std::stringstream ss;
                            ss << v;
                            _print(ss.str());
                        });
                    DJV_ASSERT(animation->isActive());
                    
                    _tickFor(std::chrono::milliseconds(300));
                    
                    animation->stop();
                    DJV_ASSERT(!animation->isActive());
                }
            }

            if (auto context = getContext().lock())
            {
                auto animation = Animation::Animation::create(context);
                animation->start(
                    1.F,
                    0.F,
                    std::chrono::milliseconds(250),
                    [this](float v)
                    {
                        std::stringstream ss;
                        ss << v;
                        _print(ss.str());
                    },
                    [this](float v)
                    {
                        std::stringstream ss;
                        ss << v;
                        _print(ss.str());
                    });
                
                _tickFor(std::chrono::milliseconds(300));
            }
        }
        
    } // namespace SystemTest
} // namespace djv

